jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -2740,6 +2742,14 @@ private String toResourceName(String resName) {
  */
 @Override
 public String toString() {
+	if (null == toStringName) {
+		toStringName = toStringImpl();
+	}
+
+	return toStringName;
+}
```

Comment:
> the second read of `toStringName` in the return statement returns null

I think that can only be described as a bug.
In any case I agree with your suggestion to use a local to avoid reading the field twice, but I would suggest separate assignment statements:
```
		localCopy = toStringImpl();
		toStringName = localCopy;
```